Index: src/org/apache/xerces/impl/xs/XMLSchemaValidator.java
===================================================================
--- src/org/apache/xerces/impl/xs/XMLSchemaValidator.java   (revision 1584342)
+++ src/org/apache/xerces/impl/xs/XMLSchemaValidator.java   (working copy)
@@ -1345,8 +1345,11 @@
 
         fIdConstraint = false;
         //reset XSDDescription
+        fXSDDescription.reset();
         fLocationPairs.clear();
         fExpandedLocationPairs.clear();
+        fUnparsedLocations.clear();
+        fGrammarBucket.reset();
 
         // cleanup id table
         fValidationState.resetIDTables();
@@ -2693,6 +2696,16 @@
                 }
             }
 
+            /*
+             * XXX: evil hack to revisit grammar pool with location information
+             */
+            if (grammar == null) {
+                grammar = (SchemaGrammar) fGrammarPool.retrieveGrammar(fXSDDescription);
+                if (grammar != null) {
+                    fXSDDescription.fLocationHints = null;
+                }
+            }
+
             if (grammar == null || fXSDDescription.fLocationHints != null) {
                 boolean toParseSchema = true;
                 if (grammar != null) {
